### PR TITLE
Fixes mentors

### PR DESCRIPTION
**Pull request by Carbonhell**
:cl: 
fix: Mentors are now fixed.
/:cl:

Do not merge, still in development. Codecheck tho

- [x] Make mentors load from text file
- [x] Make mentors load from database
- [x] Test everything, especially database shizzle
- [x] Finalize commands
- [x] Make admins able to use mentor stuff by holding the datum

### CODE REVIEW

File: code/modules/admin/topic.dm [hunk 1]
```diff
@@ -63,6 +63,7 @@
 		if (!SSticker.mode)
 			to_chat(usr, "<span class='danger'>Not until the round starts!</span>")
 			return
+		hippieTopic(href, href_list)
```
Comment (ghost) [inline code/modules/admin/topic.dm]: You've made changes to /tg/ files, please mark these changes.

**All modifications to non-hippie files should be marked.**

- Multi line changes start with `// hippie start` and end with `// hippie end`
- You can put a messages with a change if it isn't obvious, like this: `// hippie start - reason`
  - Should generally be about the reason the change was made, what it was before, or what the change is
  - Multi-line messages should start with `// hippie start` and use `/* Multi line message here */` for the message itself
**- Single line changes should have `// hippie` or `// hippie - reason`**

If you need to mirror a file, or function into a hippie-specific file, please leave behind a comment stating where it went.

```
// hippie start - Mirrored this function in <file> for <reason>
bunch of shitcode here
// hippie stop
```

Once you mirror a file, please follow the above for marking your changes, this way we know what needs to be updated when a file has been mirrored.

File: code/modules/admin/admin.dm [hunk 2]
```diff
@@ -175,6 +175,7 @@
 		body += "<A href='?_src_=holder;[HrefToken()];tdomeadmin=[REF(M)]'>Thunderdome Admin</A> | "
 		body += "<A href='?_src_=holder;[HrefToken()];tdomeobserve=[REF(M)]'>Thunderdome Observer</A> | "
 
+	body += HippiePPoptions(M)
```
Comment (ghost) [inline code/modules/admin/admin.dm]: You've made changes to /tg/ files, please mark these changes.

**All modifications to non-hippie files should be marked.**

- Multi line changes start with `// hippie start` and end with `// hippie end`
- You can put a messages with a change if it isn't obvious, like this: `// hippie start - reason`
  - Should generally be about the reason the change was made, what it was before, or what the change is
  - Multi-line messages should start with `// hippie start` and use `/* Multi line message here */` for the message itself
- Single line changes should have `// hippie` or `// hippie - reason`

If you need to mirror a file, or function into a hippie-specific file, please leave behind a comment stating where it went.

```
// hippie start - Mirrored this function in <file> for <reason>
bunch of shitcode here
// hippie stop
```

Once you mirror a file, please follow the above for marking your changes, this way we know what needs to be updated when a file has been mirrored.

File: code/modules/client/client_procs.dm [hunk 3]
```diff
@@ -82,22 +82,7 @@
...
-		return
-
-	//Mentor Follow
-	if(href_list["mentor_follow"])
-		var/mob/living/M = locate(href_list["mentor_follow"])
-
-		if(istype(M))
-			mentor_follow(M)
-
+	if(hippie_client_procs(href_list))//Mentor Msg
```
Comment (ghost) [inline code/modules/client/client_procs.dm]: You've made changes to /tg/ files, please mark these changes.

**All modifications to non-hippie files should be marked.**

- Multi line changes start with `// hippie start` and end with `// hippie end`
- You can put a messages with a change if it isn't obvious, like this: `// hippie start - reason`
  - Should generally be about the reason the change was made, what it was before, or what the change is
  - Multi-line messages should start with `// hippie start` and use `/* Multi line message here */` for the message itself
- Single line changes should have `// hippie` or `// hippie - reason`

If you need to mirror a file, or function into a hippie-specific file, please leave behind a comment stating where it went.

```
// hippie start - Mirrored this function in <file> for <reason>
bunch of shitcode here
// hippie stop
```

Once you mirror a file, please follow the above for marking your changes, this way we know what needs to be updated when a file has been mirrored.

File: code/modules/client/client_procs.dm [hunk 4]
```diff
@@ -303,10 +287,7 @@ GLOBAL_LIST_EMPTY(external_rsc_urls)
 		add_admin_verbs()
 		to_chat(src, get_message_output("memo"))
 		adminGreet()
-
-		if(!check_rights_for(src, R_ADMIN) && check_rights_for(src, R_MENTOR))
-			mentor_memo_output("Show")
-
+	hippie_mentor_holder_set()// Hippie mentor_holder set
```
Comment (ghost) [inline code/modules/client/client_procs.dm]: You've made changes to /tg/ files, please mark these changes.

**All modifications to non-hippie files should be marked.**

- Multi line changes start with `// hippie start` and end with `// hippie end`
- You can put a messages with a change if it isn't obvious, like this: `// hippie start - reason`
  - Should generally be about the reason the change was made, what it was before, or what the change is
  - Multi-line messages should start with `// hippie start` and use `/* Multi line message here */` for the message itself
- Single line changes should have `// hippie` or `// hippie - reason`

If you need to mirror a file, or function into a hippie-specific file, please leave behind a comment stating where it went.

```
// hippie start - Mirrored this function in <file> for <reason>
bunch of shitcode here
// hippie stop
```

Once you mirror a file, please follow the above for marking your changes, this way we know what needs to be updated when a file has been mirrored.

File: hippiestation/code/modules/mentor/mentor.dm [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	name = "[ckey]'s mentor datum"
+	href_token = GenerateToken()
+	//set the owner var and load commands
+	owner = GLOB.directory[ckey]
+	owner.mentor_datum = src
+	owner.add_mentor_verbs()
+	GLOB.mentors += owner
+	GLOB.mentor_datums += src
+
+/datum/mentors/proc/CheckMentorHref(href, href_list)
```
Comment (ghost) [inline hippiestation/code/modules/mentor/mentor.dm]: `CheckMentorHref` should be `checkMentorHREF`

File: hippiestation/code/modules/mentor/mentor.dm [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	var/auth = href_list["admin_token"]
```
Comment (ghost) [inline hippiestation/code/modules/mentor/mentor.dm]: `admin_token` should be `mentor_token`

File: hippiestation/code/modules/client/client_procs.dm [hunk 7]
```diff
@@ -1,4 +1,34 @@
...
+
+/client/proc/hippie_mentor_holder_set()
+	mentor_datum = GLOB.mentor_datums[ckey]
+	if(mentor_datum)
+		GLOB.mentors |= src
+		mentor_datum.owner = src
+		add_mentor_verbs()
+		mentor_memo_output("Show")
+
+/client/proc/isMentor()
```
Comment (ghost) [inline hippiestation/code/modules/client/client_procs.dm]: this should be `is_mentor()`

File: hippiestation/code/modules/admin/topic.dm [hunk 8]
```diff
@@ -13,3 +13,57 @@
...
+		return
+	var/datum/DBQuery/query_add_mentor = SSdbcore.NewQuery("INSERT INTO `[format_table_name("mentor")]` (`id`, `ckey`) VALUES (null, '[ckey]')")
+	if(!query_add_mentor.warn_execute())
+		return
+	var/datum/DBQuery/query_add_admin_log = SSdbcore.NewQuery("INSERT INTO `[format_table_name("admin_log")]` (`id` ,`datetime` ,`adminckey` ,`adminip` ,`log` ) VALUES (NULL , NOW( ) , '[usr.ckey]', '[usr.client.address]', 'Added new mentor [ckey]');")
+	if(!query_add_admin_log.warn_execute())
+		return
+	to_chat(usr, "<span class='adminnotice'>New mentor added.</span>")
+
+/datum/admins/proc/hippieRemoveMentor(ckey)
```
Comment (ghost) [inline hippiestation/code/modules/admin/topic.dm]: There is no /tg/ mentor system and thus it makes no sense to add `hippie` in front of this proc

File: hippiestation/code/modules/admin/topic.dm [hunk 9]
```diff
@@ -13,3 +13,57 @@
...
+	if(href_list["makeAntag"] == "shadowling")
+		hippie_makeShadowling(src)
+	else if(href_list["makeAntag"] == "vampire")
+		hippie_makeVampire(src)
+	else if(href_list["makementor"])
+		hippieMakeMentor(href_list["makementor"])
+	else if(href_list["removementor"])
+		hippieRemoveMentor(href_list["removementor"])
+
+/datum/admins/proc/hippieMakeMentor(ckey)
```
Comment (ghost) [inline hippiestation/code/modules/admin/topic.dm]: There is no /tg/ mentor system and thus it makes no sense to add `hippie` in front of this proc

File: code/modules/client/verbs/ooc.dm [hunk 10]
```diff
@@ -62,10 +62,10 @@
 				if(!holder.fakekey || C.holder)
 					if(check_rights_for(src, R_ADMIN))
 						to_chat(C, "<span class='adminooc'>[CONFIG_GET(flag/allow_admin_ooccolor) && prefs.ooccolor ? "<font color=[prefs.ooccolor]>" :"" ]<span class='prefix'>OOC:</span> <EM>[keyname][holder.fakekey ? "/([holder.fakekey])" : ""]:</EM> <span class='message'>[msg]</span></span></font>")
-					else
-						to_chat(C, "<font color='[GLOB.HIPPIE_MENTOR_OOC_COLOUR]'><span class='ooc'><span class='prefix'>OOC:</span> <EM>[keyname][holder.fakekey ? "/([holder.fakekey])" : ""]:</EM> <span class='message'>[msg]</span></span>")
 				else
 					to_chat(C, "<font color='[GLOB.normal_ooc_colour]'><span class='ooc'><span class='prefix'>OOC:</span> <EM>[holder.fakekey ? holder.fakekey : key]:</EM> <span class='message'>[msg]</span></span></font>")
+			else if(mentor_datum)
```
Comment (ghost) [inline code/modules/client/verbs/ooc.dm]: You've made changes to /tg/ files, please mark these changes.

**All modifications to non-hippie files should be marked.**

- Multi line changes start with `// hippie start` and end with `// hippie end`
- You can put a messages with a change if it isn't obvious, like this: `// hippie start - reason`
  - Should generally be about the reason the change was made, what it was before, or what the change is
  - Multi-line messages should start with `// hippie start` and use `/* Multi line message here */` for the message itself
- Single line changes should have `// hippie` or `// hippie - reason`

If you need to mirror a file, or function into a hippie-specific file, please leave behind a comment stating where it went.

```
// hippie start - Mirrored this function in <file> for <reason>
bunch of shitcode here
// hippie stop
```

Once you mirror a file, please follow the above for marking your changes, this way we know what needs to be updated when a file has been mirrored.

File: hippiestation/code/modules/mentor/mentor.dm (hunk 6 above)
Comment (Carbonhell) [reply hippiestation/code/modules/mentor/mentor.dm]: it actually uses the same token system of admins', not sure if i can change that without copying it somehow

**Comment by Carbonhell** [comment]
Legacy mode is confirmed to work. Href token does too. All what's missing is the sql test and i need someone to do that because i can't be assed to set up a database pls

**Comment by tortellinitony** [comment]
testing sql is for babies

**Comment by Pyko1** [comment]
yeah, real coders merge and call it a feature!

**Comment by Carbonhell** [comment]
Database works perfectly. Only thing left to fix is admins not being able to use mentor commands.

File: hippiestation/code/modules/mentor/mentorsay.dm [hunk 11]
```diff
@@ -2,7 +2,7 @@
 	set category = "Mentor"
 	set name = "Msay" //Gave this shit a shorter name so you only have to time out "msay" rather than "mentor say" to use it --NeoFite
 	set hidden = 1
-	if(!check_rights_for(src, R_MENTOR))
+	if(!is_mentor() && !check_rights_for(src, R_ADMIN,0))
```
Comment (Chnkr) [inline hippiestation/code/modules/mentor/mentorsay.dm]: Why are admins unable to use msay?

Comment (ghost) [inline hippiestation/code/modules/mentor/mentorsay.dm]: why the admin check here? Just make `is_mentor` return true if the dude is an admin

File: hippiestation/code/modules/client/client_procs.dm [hunk 12]
```diff
@@ -1,4 +1,16 @@
...
+		GLOB.mentors |= src
+		mentor_datum.owner = src
+		add_mentor_verbs()
+		mentor_memo_output("Show")
+
+/client/proc/is_mentor()
+	if(mentor_datum)
+		return TRUE
+
+/*/client/Topic/proc/HippieMentorMessage()
```
Comment (ghost) [inline hippiestation/code/modules/client/client_procs.dm]: why is this commented out? Just remove it altogether if it's not used

File: hippiestation/code/modules/client/client_procs.dm [hunk 13]
```diff
@@ -1,4 +1,16 @@
-/client/Topic/proc/HippieMentorMessage()
+/client/proc/hippie_mentor_datum_set()
```
Comment (ghost) [inline hippiestation/code/modules/client/client_procs.dm]: why is this prefixed with `hippie`

**Comment by Carbonhell** [comment]
Code ready to be fully reviewed. It only needs a final test due to issues when adminning mentors.

File: code/__DEFINES/~hippie_defines/mentor.dm [hunk 14]
```diff
@@ -0,0 +1 @@
+GLOBAL_VAR_INIT(HIPPIE_MENTOR_OOC_COLOUR, "#FF3E96")
```
Comment (tortellinitony) [inline code/__DEFINES/~hippie_defines/mentor.dm]: put this in the globalvars folder

File: config/config.txt [hunk 15]
```diff
@@ -17,7 +17,7 @@ $include hippiestation_config.txt
 # SERVERNAME tgstation
 
 ## Server SQL name: This is the name used to identify the server to the SQL DB, distinct from SERVERNAME as it must be at most 32 characters.
-# SERVERSQLNAME tgstation
+# SERVERSQLNAME hippiestation13
```
Comment (ghost) [inline config/config.txt]: No

File: config/config.txt [hunk 16]
```diff
@@ -34,10 +34,10 @@ ROUND_END_COUNTDOWN 90
 ## Comment this out if you want to use the SQL based admin system, the legacy system uses admins.txt.
 ## You need to set up your database to use the SQL based system.
 ## This flag is automatically enabled if SQL_ENABLED isn't
-ADMIN_LEGACY_SYSTEM
+#ADMIN_LEGACY_SYSTEM
```
Comment (ghost) [inline config/config.txt]: No

File: config/config.txt [hunk 17]
```diff
@@ -34,10 +34,10 @@ ROUND_END_COUNTDOWN 90
...
 
 ## Comment this out if you want to use the SQL based banning system. The legacy systems use the files in the data folder. You need to set up your database to use the SQL based system.
-BAN_LEGACY_SYSTEM
+#BAN_LEGACY_SYSTEM
```
Comment (ghost) [inline config/config.txt]: No

**Comment by Carbonhell** [comment]
There is still some leftover shit i forgot. Will fix tomorrow or monday

**Comment by Carbonhell** [comment]
I think i removed all the old admin checks by guyon/tortellini
should be ready?

**Comment by ghost** [comment]
BUT IS IT READY

**Comment by Carbonhell** [comment]
it IS GODDAMNIT

**Comment by Carbonhell** [comment]
EEEEEEEEERIIIIIIIIIIK AAAAAAAAAAAAAAAAAAAAAA

**Comment by Carbonhell** [comment]
Also readds hippie_initialize() to world.dm

**Review by ghost — APPROVED** [review]

**Comment by Carbonhell** [comment]
YEEESSSSSS

**Pull request merged**